src/lib/pattern_completeness.ml

```diff
@@ -953,4 +953,113 @@ module Make (C : Config) = struct
+                );
+              false
+          | Incomplete [] ->
+              Reporting.warn "Unreachable pattern match type at" (shrink_loc keyword l)
+                ("The following type is unreachable: " ^ (string_of_typ typ |> Util.yellow |> Util.clear));
+              false
+          | Complete cinfo ->
+              List.iter
+                (fun (idx, _) ->
+                  if IntSet.mem idx.num cinfo.redundant then
+                    Reporting.warn "Redundant case" idx.loc "This match case is never used"
+                )
+                (rows_to_list matrix);
+              true
+          | Completeness_unknown -> false
+        in
+        all_complete && complete
+      )
+      true typ2matrixs
+
+  let is_complete_mapcls_wildcarded ?(keyword = "match") l ctx mapcls typl typr =
+    let have_guard = false in
+    let rec lpat_of_mapcl mapcl left =
+      (* We don't consider guarded cases *)
+      let pat_of_mpexp = function
+        | MPat_aux (mpexp_aux, l) -> (
+            match mpexp_aux with MPat_when (mpat, exp) -> None | MPat_pat mpat -> Some (pat_of_mpat mpat)
+          )
+      in
+      let rec pexp_to_pat = function
+        | Pat_aux (Pat_exp ((P_aux (_, (l, _)) as pat), _), _) -> Some pat
+        | Pat_aux (Pat_when _, _) -> None
+      in
+      match mapcl with
+      | MCL_aux (mapcl_aux, _) ->
+          if left then (
+            match mapcl_aux with
+            | MCL_bidir (mpexp, _) -> pat_of_mpexp mpexp
+            | MCL_forwards pexp -> pexp_to_pat pexp
+            | MCL_backwards pexp -> None
+          )
+          else (
+            match mapcl_aux with
+            | MCL_bidir (_, mpexp) -> pat_of_mpexp mpexp
+            | MCL_forwards pexp -> None
+            | MCL_backwards pexp -> pexp_to_pat pexp
+          )
+    in
+    let rec opt_cases_to_pats from have_guard = function
+      | [] -> (have_guard, [])
+      | Some (P_aux (_, (l, _)) as pat) :: cases ->
+          let pat, from = number_pat from pat in
+          let have_guard, pats = opt_cases_to_pats from have_guard cases in
+          (have_guard, (l, pat) :: pats)
+      | _ :: cases -> opt_cases_to_pats from true cases
+    in
+    (* TODO: handle unsupported syntax *)
+    try
+      (* prepare left *)
+      let lpats = List.map (fun mapcl -> lpat_of_mapcl mapcl true) mapcls in
+      let _, lpats = opt_cases_to_pats 0 have_guard lpats in
+      let lmatrix =
+        Rows (List.mapi (fun i (l, pat) -> ({ loc = l; num = i }, Columns [generalize ctx (Some typl) pat])) lpats)
+      in
+      (* prepare right *)
+      let rpats = List.map (fun mapcl -> lpat_of_mapcl mapcl false) mapcls in
+      let _, rpats = opt_cases_to_pats 0 have_guard rpats in
+      let rmatrix =
+        Rows (List.mapi (fun i (l, pat) -> ({ loc = l; num = i }, Columns [generalize ctx (Some typr) pat])) rpats)
+      in
+      is_complete_matrixs ~keyword:"mapping" l ctx have_guard [(typl, lmatrix); (typr, rmatrix)]
+    with _ -> false
```

Comment:
Some syntaxes are not supported in pat, but are allowed by mpat

example: _test/c/poly_mapping2.sail_

```
mapping test2m : bits(3) <-> bits(5) = { v <-> zero_int_bits(2) : bits(2) @ v }
```

When encountering this kind of syntax, an exception may be thrown, so here fallback to false.